components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2239,17 +2239,88 @@ where
                     if self.fsm.batch_req_builder.request.is_some() {
                         self.propose_batch_raft_command(true);
                     }
+                    if self.propose_locks_before_transfer_leader() {
+                        // If some pessimistic locks are just proposed, we propose another
+                        // TransferLeader command instead of transferring leader immediately.
 
-                    self.fsm.peer.transfer_leader(&from);
+                        let mut cmd = new_admin_request(
+                            self.fsm.peer.region().get_id(),
+                            self.fsm.peer.peer.clone(),
+                        );
+                        cmd.mut_header()
+                            .set_region_epoch(self.region().get_region_epoch().clone());
+                        cmd.mut_admin_request()
+                            .set_cmd_type(AdminCmdType::TransferLeader);
+                        cmd.mut_admin_request().mut_transfer_leader().set_peer(from);
+                        self.propose_raft_command(
+                            cmd,
+                            Callback::None,
+                            DiskFullOpt::AllowedOnAlmostFull,
+                        );
+                    } else {
+                        self.fsm.peer.transfer_leader(&from);
+                    }
                 }
             }
         } else {
             self.fsm
                 .peer
-                .execute_transfer_leader(&mut self.ctx, msg, peer_disk_usage);
+                .execute_transfer_leader(&mut self.ctx, msg.get_from(), peer_disk_usage);
         }
     }
 
+    // Returns whether we should propose another TransferLeader command. This is for:
+    // 1. Considering the amount of pessimistic locks can be big, it can reduce
+    // unavailable time caused by waiting for the transferree catching up logs.
+    // 2. Make transferring leader strictly after write commands that executes
+    // before proposing the locks, preventing unexpected lock loss.
+    fn propose_locks_before_transfer_leader(&mut self) -> bool {
+        // 1. Disable in-memory pessimistic locks.
```

Comment:
I just find it incorrect to take away the locks here. After the locks are taken away, write requests will fail to read the locks here, while the locks are also missing from the underlying rocksdb.

I'll fix it and add tests to cover the case.